src/EditorFeatures/Core.Wpf/InlineErrors/InlineErrorTaggerProvider.cs

```diff
@@ -0,0 +1,118 @@
+using Microsoft.VisualStudio.Text.Tagging;
+using Microsoft.VisualStudio.Utilities;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Editor.InlineErrors
+{
+    [Export(typeof(ITaggerProvider))]
+    [ContentType(ContentTypeNames.RoslynContentType)]
+    [ContentType(ContentTypeNames.XamlContentType)]
+    [TagType(typeof(InlineErrorTag))]
+    internal class InlineErrorTaggerProvider : AbstractDiagnosticsAdornmentTaggerProvider<InlineErrorTag>
+    {
+        private readonly IEditorFormatMap _editorFormatMap;
+        protected sealed override IEnumerable<PerLanguageOption2<bool>> PerLanguageOptions => SpecializedCollections.SingletonEnumerable(InlineErrorsOptions.EnableInlineDiagnostics);
+
+        protected internal override bool IsEnabled => true;
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public InlineErrorTaggerProvider(
+            IThreadingContext threadingContext,
+            IEditorFormatMapService editorFormatMapService,
+            IDiagnosticService diagnosticService,
+            IAsynchronousOperationListenerProvider listenerProvider)
+            : base(threadingContext, diagnosticService, listenerProvider)
+        {
+            _editorFormatMap = editorFormatMapService.GetEditorFormatMap("text");
+        }
+
+        protected override ITaggerEventSource CreateEventSource(ITextView textViewOpt, ITextBuffer subjectBuffer)
+        {
+            return TaggerEventSources.Compose(
+                TaggerEventSources.OnDocumentActiveContextChanged(subjectBuffer),
+                TaggerEventSources.OnWorkspaceRegistrationChanged(subjectBuffer),
+                TaggerEventSources.OnDiagnosticsChanged(subjectBuffer, DiagnosticService),
+                TaggerEventSources.OnOptionChanged(subjectBuffer, InlineErrorsOptions.EnableInlineDiagnostics),
+                TaggerEventSources.OnOptionChanged(subjectBuffer, InlineErrorsOptions.LocationOption));
+        }
+
+        protected internal override bool IncludeDiagnostic(DiagnosticData diagnostic)
+        {
+            return
+                diagnostic.Severity is DiagnosticSeverity.Warning or DiagnosticSeverity.Error &&
+                !string.IsNullOrWhiteSpace(diagnostic.Message) && !diagnostic.IsSuppressed;
+        }
+
+        /// <summary>
+        /// Creates the InlineErrorTag with the error distinction
+        /// </summary>
+        protected override InlineErrorTag? CreateTag(Workspace workspace, DiagnosticData diagnostic)
+        {
+            Debug.Assert(!string.IsNullOrWhiteSpace(diagnostic.Message));
+            var errorType = GetErrorTypeFromDiagnostic(diagnostic);
+            if (errorType is null)
+            {
+                return null;
+            }
+
+            RoslynDebug.AssertNotNull(diagnostic.DocumentId);
+            var document = workspace.CurrentSolution.GetRequiredDocument(diagnostic.DocumentId);
```

Comment:
this is dangerous. t he doc might be gone at this point.  i would get the doc, and just bail if you get null.